tools/monitoring/config_test.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"reflect"
+	"testing"
+)
+
+const (
+	sampleYaml = `spec:
+  - error-pattern: 'Something went wrong: starting e2e cluster: error creating cluster'
+    hint: 'Check gcp status'
+    alerts:
+      - job-name-regex: '^pull.*'
+        occurrences: 2
+        jobs-affected: 1
+        prs-affected: 2
+        period: 1440 # 1440 minutes = 24 hours
+      - job-name-regex: '.*'
+        occurrences: 5
+        jobs-affected: 2
+        prs-affected: 1 # for non-pull jobs, we don't care about the number of prs affected, so we set the number to 1, which will basically make this particular condition always true
+        period: 1440
+
+  - error-pattern: 'sample*error2'
+    hint: 'hint_for_pattern_2'
+    alerts:
+      - job-name-regex: '^pull.*'
+        occurrences: 20
+        jobs-affected: 10
+        prs-affected: 20
+        period: 60 # 1440 minutes = 24 hours
+      - job-name-regex: '.*'
+        occurrences: 50
+        jobs-affected: 20
+        prs-affected: 10 # for non-pull jobs, we don't care about the number of prs affected, so we set the number to 1, which will basically make this particular condition always true
+        period: 60`
+)
+
+var sampleConfig = Config{
+	Spec: []patternSpec{
+		{
+			ErrorPattern: "Something went wrong: starting e2e cluster: error creating cluster",
+			Hint:         "Check gcp status",
+			Alerts: []alertCondition{
+				{"^pull.*",
+					2,
+					1,
+					2,
+					1440},
+				{".*",
+					5,
+					2,
+					1,
+					1440},
+			},
+		},
+
+		{
+			"sample*error2",
+			"hint_for_pattern_2",
+			[]alertCondition{
+				{"^pull.*",
+					20,
+					10,
+					20,
+					60,
+				},
+				{".*",
+					50,
+					20,
+					10,
+					60,
+				},
+			},
+		},
+	},
+}
+
+func TestCollectErrorPatterns(t *testing.T) {
+	conf, err := newConfig([]byte(sampleYaml))
+	if err != nil {
+		t.Errorf("cannot parse yaml: %v", err)
+	}
+
+	type args struct {
```

Comment:
do we need this struct? can we directly have f *Config in the tests struct?

Reply:
Having this struct in all unit tests brings consistency. It's the default way intellij constructs unit tests

Reply:
that is an IDE generalization and does not mean we should do that. It makes sense in most of the cases, but in this case, it adds an unnecessary level of indirection. We can keep it if you want to be consistent across the tests.